File: backend/src/services/reportExport.ts
import PDFDocument from "pdfkit";
import { Document, Packer, Paragraph, TextRun, HeadingLevel, AlignmentType } from "docx";
import type { StudentRow } from "./db";
import { explainApplicationStatus } from "./applicationStatus";
import {
  PREFERRED_COUNTRY_ANY,
  ACADEMIC_BACKGROUND_LABELS,
  MIGRATION_INTENT_LABELS,
  type AcademicBackground,
  type MigrationIntent,
} from "./leadScoring";

interface ReportData {
  name: string;
  email: string | null;
  phone: string | null;
  referenceCode: string | null;
  generatedAt: string;
  profile: { label: string; value: string }[];
  recommendedCountries: string;
  aiResponse: string;
  nextSteps: string[];
  applicationStatus: { label: string; description: string };
}

// Deliberately excludes internal CRM-only fields (Hot/Warm/Cold score,
// counsellor notes, assignment, follow-up data) — this document is the
// student's own copy of their recommendation, not a CRM record.
function buildReportData(row: StudentRow): ReportData {
  const statusInfo = explainApplicationStatus(row.application_status);
  return {
    name: row.name,
    email: row.email,
    phone: row.phone,
    referenceCode: row.reference_code,
    generatedAt: new Date().toLocaleDateString("en-US", { year: "numeric", month: "long", day: "numeric" }),
    profile: [
      { label: "GPA", value: `${row.gpa} / 4.0` },
      { label: "IELTS", value: `${row.ielts} / 9.0` },
      { label: "Annual Budget", value: `$${row.budget.toLocaleString()} USD` },
      { label: "Study Gap", value: `${row.gap} year(s)` },
      {
        label: "Academic Background",
        value: row.academic_background
          ? (ACADEMIC_BACKGROUND_LABELS[row.academic_background as AcademicBackground] ?? row.academic_background)
          : "—",
      },
      { label: "Career Goals", value: row.career_goals || "—" },
      {
        label: "Migration Intent",
        value: row.migration_intent
          ? (MIGRATION_INTENT_LABELS[row.migration_intent as MigrationIntent] ?? row.migration_intent)
          : "—",
      },
      {
        label: "Preferred Country",
        value:
          row.preferred_country && row.preferred_country !== PREFERRED_COUNTRY_ANY
            ? row.preferred_country
            : "Any / Let AI decide",
      },
    ],
    recommendedCountries: row.countries || "—",
    aiResponse: row.ai_response || "(No recommendation on file yet.)",
    nextSteps: row.next_steps ? (JSON.parse(row.next_steps) as string[]) : [],
    applicationStatus: { label: statusInfo.label, description: statusInfo.description },
  };
}

export function generateReportPdf(row: StudentRow): Promise<Buffer> {
  const data = buildReportData(row);

  return new Promise((resolve, reject) => {
    const doc = new PDFDocument({ margin: 50 });
    const chunks: Buffer[] = [];
    doc.on("data", (chunk) => chunks.push(chunk));
    doc.on("end", () => resolve(Buffer.concat(chunks)));
    doc.on("error", reject);

    doc.fontSize(20).font("Helvetica-Bold").fillColor("#1a5f7a").text("AIEC Global");
    doc.fontSize(13).font("Helvetica").fillColor("#555").text("Study Abroad Recommendation Report");
    doc.moveDown(0.3);
    doc
      .fontSize(9)
      .fillColor("#999")
      .text(`Generated ${data.generatedAt}${data.referenceCode ? `  ·  Reference: ${data.referenceCode}` : ""}`);
    doc.moveDown(1);
    doc.fillColor("#000");

    doc.fontSize(14).font("Helvetica-Bold").text(data.name);
    const contactLine = [data.email, data.phone].filter(Boolean).join("  ·  ");
    if (contactLine) doc.fontSize(10).font("Helvetica").fillColor("#555").text(contactLine);
    doc.fillColor("#000");
    doc.moveDown(1);

    sectionHeading(doc, "Your Profile");
    data.profile.forEach(({ label, value }) => {
      doc.fontSize(10).font("Helvetica-Bold").text(`${label}: `, { continued: true }).font("Helvetica").text(value);
    });
    doc.moveDown(1);

    sectionHeading(doc, "Recommended Countries");
    doc.fontSize(10).font("Helvetica").text(data.recommendedCountries);
    doc.moveDown(1);

    sectionHeading(doc, "AI Recommendation");
    doc.fontSize(10).font("Helvetica").text(data.aiResponse, { align: "left" });
    doc.moveDown(1);

    if (data.nextSteps.length > 0) {
      sectionHeading(doc, "Next Steps");
      data.nextSteps.forEach((step, i) => {
        doc.fontSize(10).font("Helvetica").text(`${i + 1}. ${step}`);
      });
      doc.moveDown(1);
    }

    sectionHeading(doc, "Current Application Status");
    doc
      .fontSize(10)
      .font("Helvetica-Bold")
      .text(data.applicationStatus.label, { continued: true })
      .font("Helvetica")
      .text(` — ${data.applicationStatus.description}`);

    doc.moveDown(2);
    doc.fontSize(8).fillColor("#999").text("Questions about this report? Contact your AIEC Global counsellor.");

    doc.end();
  });
}

function sectionHeading(doc: PDFKit.PDFDocument, text: string): void {
  doc.fontSize(12).font("Helvetica-Bold").fillColor("#1a5f7a").text(text);
  doc.fillColor("#000");
  doc.moveDown(0.3);
}

export async function generateReportDocx(row: StudentRow): Promise<Buffer> {
  const data = buildReportData(row);

  const heading = (text: string) =>
    new Paragraph({ heading: HeadingLevel.HEADING_2, spacing: { before: 240, after: 120 }, children: [new TextRun({ text, color: "1a5f7a" })] });

  const doc = new Document({
    sections: [
      {
        children: [
          new Paragraph({
            children: [new TextRun({ text: "AIEC Global", bold: true, size: 40, color: "1a5f7a" })],
          }),
          new Paragraph({
            children: [new TextRun({ text: "Study Abroad Recommendation Report", size: 26, color: "555555" })],
            spacing: { after: 80 },
          }),
          new Paragraph({
            children: [
              new TextRun({
                text: `Generated ${data.generatedAt}${data.referenceCode ? `  ·  Reference: ${data.referenceCode}` : ""}`,
                size: 18,
                color: "999999",
              }),
            ],
            spacing: { after: 200 },
          }),
          new Paragraph({ children: [new TextRun({ text: data.name, bold: true, size: 28 })] }),
          ...(data.email || data.phone
            ? [
                new Paragraph({
                  children: [
                    new TextRun({ text: [data.email, data.phone].filter(Boolean).join("  ·  "), size: 20, color: "555555" }),
                  ],
                  spacing: { after: 200 },
                }),
              ]
            : []),

          heading("Your Profile"),
          ...data.profile.map(
            ({ label, value }) =>
              new Paragraph({
                children: [new TextRun({ text: `${label}: `, bold: true }), new TextRun({ text: value })],
              }),
          ),

          heading("Recommended Countries"),
          new Paragraph({ text: data.recommendedCountries }),

          heading("AI Recommendation"),
          ...data.aiResponse.split("\n").map((line) => new Paragraph({ text: line })),

          ...(data.nextSteps.length > 0
            ? [heading("Next Steps"), ...data.nextSteps.map((step, i) => new Paragraph({ text: `${i + 1}. ${step}` }))]
            : []),

          heading("Current Application Status"),
          new Paragraph({
            children: [
              new TextRun({ text: data.applicationStatus.label, bold: true }),
              new TextRun({ text: ` — ${data.applicationStatus.description}` }),
            ],
          }),

          new Paragraph({
            spacing: { before: 400 },
            alignment: AlignmentType.LEFT,
            children: [
              new TextRun({ text: "Questions about this report? Contact your AIEC Global counsellor.", size: 16, color: "999999" }),
            ],
          }),
        ],
      },
    ],
  });

  return Packer.toBuffer(doc);
}

File: backend-py/app/services/report_export.py
import io
import json
from datetime import datetime
from typing import Any

from docx import Document
from docx.shared import Pt, RGBColor
from reportlab.lib.pagesizes import letter
from reportlab.pdfgen import canvas

from app.services.application_status import explain_application_status
from app.services.lead_scoring import (
    ACADEMIC_BACKGROUND_LABELS,
    MIGRATION_INTENT_LABELS,
    PREFERRED_COUNTRY_ANY,
)

_BRAND_COLOR = "1a5f7a"


def _build_report_data(row: Any) -> dict:
    """Deliberately excludes internal CRM-only fields (Hot/Warm/Cold score,
    counsellor notes, assignment, follow-up data) — this document is the
    student's own copy of their recommendation, not a CRM record."""
    status_info = explain_application_status(row["application_status"])
    academic_background = row["academic_background"]
    migration_intent = row["migration_intent"]
    preferred_country = row["preferred_country"]

    return {
        "name": row["name"],
        "email": row["email"],
        "phone": row["phone"],
        "referenceCode": row["reference_code"],
        "generatedAt": datetime.now().strftime("%B %d, %Y"),
        "profile": [
            {"label": "GPA", "value": f"{row['gpa']} / 4.0"},
            {"label": "IELTS", "value": f"{row['ielts']} / 9.0"},
            {"label": "Annual Budget", "value": f"${row['budget']:,.0f} USD"},
            {"label": "Study Gap", "value": f"{row['gap']} year(s)"},
            {
                "label": "Academic Background",
                "value": ACADEMIC_BACKGROUND_LABELS.get(academic_background, academic_background) if academic_background else "—",
            },
            {"label": "Career Goals", "value": row["career_goals"] or "—"},
            {
                "label": "Migration Intent",
                "value": MIGRATION_INTENT_LABELS.get(migration_intent, migration_intent) if migration_intent else "—",
            },
            {
                "label": "Preferred Country",
                "value": preferred_country if preferred_country and preferred_country != PREFERRED_COUNTRY_ANY else "Any / Let AI decide",
            },
        ],
        "recommendedCountries": row["countries"] or "—",
        "aiResponse": row["ai_response"] or "(No recommendation on file yet.)",
        "nextSteps": json.loads(row["next_steps"]) if row["next_steps"] else [],
        "applicationStatus": {"label": status_info.label, "description": status_info.description},
    }


def _wrap_text(text: str, width_chars: int) -> list[str]:
    lines: list[str] = []
    for paragraph in text.split("\n"):
        words = paragraph.split(" ")
        current = ""
        for word in words:
            candidate = f"{current} {word}".strip()
            if len(candidate) > width_chars and current:
                lines.append(current)
                current = word
            else:
                current = candidate
        lines.append(current)
    return lines


def generate_report_pdf(row: Any) -> bytes:
    data = _build_report_data(row)
    buffer = io.BytesIO()
    c = canvas.Canvas(buffer, pagesize=letter)
    width, height = letter
    margin = 50
    y = height - margin

    def heading(text: str) -> None:
        nonlocal y
        c.setFont("Helvetica-Bold", 12)
        c.setFillColorRGB(0x1A / 255, 0x5F / 255, 0x7A / 255)
        c.drawString(margin, y, text)
        c.setFillColorRGB(0, 0, 0)
        y -= 18

    def line(text: str, font: str = "Helvetica", size: int = 10, gap: int = 14) -> None:
        nonlocal y
        c.setFont(font, size)
        for wrapped in _wrap_text(text, 95):
            if y < margin + 40:
                c.showPage()
                y = height - margin
                c.setFont(font, size)
            c.drawString(margin, y, wrapped)
            y -= gap

    c.setFont("Helvetica-Bold", 20)
    c.setFillColorRGB(0x1A / 255, 0x5F / 255, 0x7A / 255)
    c.drawString(margin, y, "AIEC Global")
    y -= 20
    c.setFont("Helvetica", 13)
    c.setFillColorRGB(0.33, 0.33, 0.33)
    c.drawString(margin, y, "Study Abroad Recommendation Report")
    y -= 16
    c.setFont("Helvetica", 9)
    c.setFillColorRGB(0.6, 0.6, 0.6)
    ref_suffix = f"  ·  Reference: {data['referenceCode']}" if data["referenceCode"] else ""
    c.drawString(margin, y, f"Generated {data['generatedAt']}{ref_suffix}")
    y -= 24
    c.setFillColorRGB(0, 0, 0)

    c.setFont("Helvetica-Bold", 14)
    c.drawString(margin, y, data["name"])
    y -= 16
    contact_line = "  ·  ".join([v for v in [data["email"], data["phone"]] if v])
    if contact_line:
        c.setFont("Helvetica", 10)
        c.setFillColorRGB(0.33, 0.33, 0.33)
        c.drawString(margin, y, contact_line)
        y -= 14
    c.setFillColorRGB(0, 0, 0)
    y -= 10

    heading("Your Profile")
    for item in data["profile"]:
        line(f"{item['label']}: {item['value']}")
    y -= 10

    heading("Recommended Countries")
    line(data["recommendedCountries"])
    y -= 10

    heading("AI Recommendation")
    line(data["aiResponse"])
    y -= 10

    if data["nextSteps"]:
        heading("Next Steps")
        for i, step in enumerate(data["nextSteps"], start=1):
            line(f"{i}. {step}")
        y -= 10

    heading("Current Application Status")
    line(f"{data['applicationStatus']['label']} — {data['applicationStatus']['description']}")

    y -= 20
    c.setFont("Helvetica", 8)
    c.setFillColorRGB(0.6, 0.6, 0.6)
    c.drawString(margin, y, "Questions about this report? Contact your AIEC Global counsellor.")

    c.save()
    return buffer.getvalue()


def generate_report_docx(row: Any) -> bytes:
    data = _build_report_data(row)
    doc = Document()
    brand = RGBColor(0x1A, 0x5F, 0x7A)
    grey = RGBColor(0x55, 0x55, 0x55)
    light_grey = RGBColor(0x99, 0x99, 0x99)

    def heading(text: str) -> None:
        p = doc.add_paragraph()
        p.paragraph_format.space_before = Pt(12)
        p.paragraph_format.space_after = Pt(6)
        run = p.add_run(text)
        run.bold = True
        run.font.size = Pt(14)
        run.font.color.rgb = brand

    title = doc.add_paragraph()
    run = title.add_run("AIEC Global")
    run.bold = True
    run.font.size = Pt(20)
    run.font.color.rgb = brand

    subtitle = doc.add_paragraph()
    run = subtitle.add_run("Study Abroad Recommendation Report")
    run.font.size = Pt(13)
    run.font.color.rgb = grey

    meta = doc.add_paragraph()
    ref_suffix = f"  ·  Reference: {data['referenceCode']}" if data["referenceCode"] else ""
    run = meta.add_run(f"Generated {data['generatedAt']}{ref_suffix}")
    run.font.size = Pt(9)
    run.font.color.rgb = light_grey

    name_p = doc.add_paragraph()
    run = name_p.add_run(data["name"])
    run.bold = True
    run.font.size = Pt(14)

    contact_line = "  ·  ".join([v for v in [data["email"], data["phone"]] if v])
    if contact_line:
        contact_p = doc.add_paragraph()
        run = contact_p.add_run(contact_line)
        run.font.size = Pt(10)
        run.font.color.rgb = grey

    heading("Your Profile")
    for item in data["profile"]:
        p = doc.add_paragraph()
        r1 = p.add_run(f"{item['label']}: ")
        r1.bold = True
        p.add_run(item["value"])

    heading("Recommended Countries")
    doc.add_paragraph(data["recommendedCountries"])

    heading("AI Recommendation")
    for para in data["aiResponse"].split("\n"):
        doc.add_paragraph(para)

    if data["nextSteps"]:
        heading("Next Steps")
        for i, step in enumerate(data["nextSteps"], start=1):
            doc.add_paragraph(f"{i}. {step}")

    heading("Current Application Status")
    p = doc.add_paragraph()
    r1 = p.add_run(data["applicationStatus"]["label"])
    r1.bold = True
    p.add_run(f" — {data['applicationStatus']['description']}")

    footer = doc.add_paragraph()
    footer.paragraph_format.space_before = Pt(20)
    run = footer.add_run("Questions about this report? Contact your AIEC Global counsellor.")
    run.font.size = Pt(8)
    run.font.color.rgb = light_grey

    buffer = io.BytesIO()
    doc.save(buffer)
    return buffer.getvalue()
